core/src/avm2/frame.rs

```diff
@@ -0,0 +1,46 @@
+use std::cmp::min;
+use std::ops::Deref;
+
+#[derive(Debug)]
+pub struct StackFrame<'a, T> {
+    stack: &'a mut Vec<T>,
+    depth: usize,
+    max: usize,
+}
+
+impl<'a, T> StackFrame<'a, T> {
+    pub fn new(stack: &'a mut Vec<T>, depth: usize, max: usize) -> Self {
+        Self { stack, depth, max }
+    }
+
+    pub fn clear(&mut self) {
+        self.stack.truncate(self.depth);
+    }
+
+    pub fn push(&mut self, value: T) {
+        if self.len() > self.max {
+            log::warn!("Avm2::StackFrame::push overflow");
+            return;
+        }
+        self.stack.push(value);
+    }
+
+    pub fn pop(&mut self) -> Option<T> {
+        if !self.is_empty() {
+            self.stack.pop()
+        } else {
+            None
+        }
```

Comment:
Can this just be `self.stack.pop()` because it returns `None` anyway when empty?

Reply:
This code is a bit confusing, but the `is_empty` is actually checking if the current frame is empty, not the whole stack, so there might still be values before the frame. If we just did `self.stack.pop()`, we could potentially pop a value from outside of the stack frame.